src/zephyr_coap_req.c

```diff
@@ -761,6 +762,59 @@ static void golioth_coap_reqs_cancel_all_with_reason(struct golioth_client *clie
     }
 }
 
+static int __golioth_coap_req_find_and_cancel(golioth_coap_request_msg_t *cancel_req,
+                                              struct golioth_client *client)
+{
+    struct golioth_coap_req *req, *next;
+
+    SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&client->coap_reqs, req, next, node)
+    {
+        if (req->user_data == cancel_req)
+        {
+            int err;
+
+            uint8_t coap_token[COAP_TOKEN_MAX_LEN];
+            size_t coap_token_len = coap_header_get_token(&req->request, coap_token);
+            int coap_content_format = coap_get_option_int(&req->request, COAP_OPTION_ACCEPT);
+
+            if (coap_token_len < 0)
+            {
+                LOG_ERR("Unable to get coap token from request. Got length: %d", coap_token_len);
+                return coap_content_format;
+            }
+
+            if (coap_content_format < 0)
+            {
+                LOG_ERR("Unable to get coap content format from request: %d", coap_content_format);
+                return coap_content_format;
+            }
+
+            golioth_coap_request_msg_t *req_msg = req->user_data;
+
+            golioth_coap_req_cancel(req);
+            err = golioth_coap_client_observe_release_async(client,
```

Comment:
This feels a little wonky. `golioth_coap_request_msg_t` can be used for all requests, not just observations, right? So it's non-obvious looking at this function signature or the function comment that we will attempt to release an observation unconditionally while cancelling the request.